janitor/functions/complete.py

```diff
@@ -212,112 +212,141 @@ def _computations_complete(
-    trimmed = None
-    indicator = None
-    # iteration used here, instead of assign (which is also a for loop),
-    # to cater for scenarios where the column_name is not a string
-    # assign only works with keys that are strings
-    for column_name, value in fill_value.items():
-        uniques[column_name] = value
-    df = pd.concat([df, uniques], sort=False, copy=False, axis="index")
-    if sort:
-        df = df.sort_values(column_checker)
-    df.index = range(len(df))
+
+    if fill_value is not None:
+        if is_scalar(fill_value):
+            # faster when fillna operates on a Series basis
+            fill_value = {
+                col: fill_value for col in columns if df[col].hasnans
+            }
+        if explicit:
+            df = df.fillna(fill_value, downcast="infer")
+        else:
+            # keep only columns that are not part of column_checker
+            # IOW, we are excluding columns that were not used
+            # to generate the combinations
+            fill_value = {
+                col: value
+                for col, value in fill_value.items()
+                if col not in column_checker
+            }
+            if fill_value:
+                # when explicit is False
+                # use the indicator parameter to identify rows
+                # for `left_only`, and fill the relevant columns in fill_value
+                # with the associated value.
+                boolean_filter = df.loc[:, indicator] == "left_only"
+                df = df.drop(columns=indicator)
+                # iteration used here,
+                # instead of assign (which is also a for loop),
+                # to cater for scenarios where the column_name is not a string
+                # assign only works with keys that are strings
+                for column_name, value in fill_value.items():
+                    # for categorical dtypes, set the categories first
+                    if is_categorical_dtype(df[column_name]):
+                        df[column_name] = df[column_name].cat.add_categories(
+                            [value]
+                        )
+                    df.loc[boolean_filter, column_name] = value
+
+    if not df.columns.equals(columns):
+        return df.reindex(columns=columns)
     return df
 
 
-def _generic_complete(df: pd.DataFrame, columns: list, all_strings: bool):
+def _generic_complete(
+    df: pd.DataFrame, columns: list, all_strings: bool, sort: bool
+):
     """
     Generate cartesian product for `_computations_complete`.
 
-    Returns a Series or DataFrame, with no duplicates.
+    Returns a DataFrame, with no duplicates.
     """
     if all_strings:
-        uniques = {col: df[col].unique() for col in columns}
+        if sort:
+            uniques = {}
+            for col in columns:
+                column = extract_array(df[col], extract_numpy=True)
+                _, column = pd.factorize(column, sort=sort)
```

Comment:
Is there any value to extracting the array ourselves here rather than leaving it to pandas?
`pd.factorize` / `pd.Series.factorize` seems to already [do this internally](https://github.com/pandas-dev/pandas/blob/06d230151e6f18fdb8139d09abf539867a8cd481/pandas/core/algorithms.py#L725).

Reply:
Only if it is already an array; any pandas object is converted to an Index, except Categorical, which returns a categorical.

```py
In [2]: box = ['b', 'b', 'a', 'c', 'b']

In [3]: box = pd.Series(box)

In [4]: pd.factorize(box)
Out[4]: (array([0, 0, 1, 2, 0]), Index(['b', 'a', 'c'], dtype='object'))

In [5]: pd.factorize(box)[-1]
Out[5]: Index(['b', 'a', 'c'], dtype='object')
```